README.md
# PySOGS (Session Community Server)

PySOGS is the reference implementation of a Session Community server (formerly known as a Session
Open Group). PySOGS is used to run the official Session Communities, and is the officially supported
Session Community server.  New features added to Session Communities are developed here in lockstep
with the support added to the Session clients.

## Installation

For most servers we provide and recommend using the .deb installation method on a VPS running Ubuntu
20.04 (or newer), or Debian 11 (or newer).

Alternatively advanced users and developers interested in running or working on the latest
development code may prefer to run directly from the repository using uwsgi.

While both methods are supported modes of operation, the latter requires more configuration and
maintenance and requires some experience with running Python code and web applications.

[Debian/Ubuntu Package Install](install-debs.md)

[Manual Installation](install-uwsgi.md)

## Administration

For how to administer a running PySOGS see [SOGS Administration](administration.md).

## License

Copyright (c) 2021-2022 The Oxen Project

PySOGS is licensed under the [GNU General Public License (GPL) v3](LICENSE.txt), or (at your option)
any later version.


debian/changelog
session-pysogs (0.3.7-2) sid; urgency=medium

  [ Jason Rhinelander ]
  * Rediff patches

  [ 平民·寻梦（Pingmin Fenlly Liu） ]
  * Fix the nginx proxy_pass pattern

 -- Jason Rhinelander <jason@imaginary.ca>  Wed, 21 Jun 2023 11:08:13 -0300

session-pysogs (0.3.7-1) sid; urgency=medium

  [ Jason Rhinelander ]
  * bump dev version
  * Move tomli to setup_requires

  [ zcyph ]
  * switch qrencode to qrcode

  [ Jason Rhinelander ]
  * Update qrencode->qrcode in package dependencies
  * Better room modification CLI arguments; allow --description modification
  * Improve errors
  * Improve `--verbose` description
  * Fix room deletion when a room has reactions
  * dev version bump

 -- Jason Rhinelander <jason@imaginary.ca>  Mon, 16 Jan 2023 16:32:46 -0400

session-pysogs (0.3.6-2) sid; urgency=medium

  * Add ignore-new for gbp pq
  * Move tomli to setup_requires

 -- Jason Rhinelander <jason@imaginary.ca>  Thu, 15 Dec 2022 11:55:55 -0400

session-pysogs (0.3.6-1) sid; urgency=medium

  [ Jason Rhinelander ]
  * Remove expired DMs during cleanup

  [ Ian Macdonald ]
  * Log only HTTP errors for much less verbose logging.

  [ Jeff Becker ]
  * add endpoint to delete own DMs

  [ Jason Rhinelander ]
  * Require either sender or recipient for deletion
  * Fix test suite use of `r.json is None`
  * docstring formatting/rewording
  * 0.3.6 release

 -- Jason Rhinelander <jason@imaginary.ca>  Fri, 02 Dec 2022 13:39:53 -0400

session-pysogs (0.3.5-1) sid; urgency=medium

  * Change "open groups" -> "communities" in readme
  * Keep 2021 in copyright range
  * Stable version bump

 -- Jason Rhinelander <jason@imaginary.ca>  Tue, 01 Nov 2022 19:51:42 -0300

session-pysogs (0.3.5~rc1-2) sid; urgency=medium

  * Fix `python3 -msogs` => `sogs` for deb

 -- Jason Rhinelander <jason@imaginary.ca>  Thu, 20 Oct 2022 18:11:06 -0300

session-pysogs (0.3.5~rc1-1) sid; urgency=medium

  [ Jason Rhinelander ]
  * Add tests for global mods/admins
  * Check for blinding when de-modding/unbanning
  * Fix "stuck moderator" problem
  * Fix unblinded legacy auth
  * Add room reaction use to CLI room display
  * Add sid CI workaround

  [ Ian Macdonald ]
  * Allow rejection of postings written in a particular alphabet.

  [ Jason Rhinelander ]
  * Add missing ngnix proxied headers to sample config
  * filter responses, overhaul filter configuration
  * Add seqno_creation to be able to skip irrelevant deletions
  * Defer post parsing until needed
  * upgrade & pg import: Add/handle seqno_creation
  * Fix postgres docker image version
  * Add some other missing pull: always

  [ Thomas Winget ]
  * add --set-perms to pysogs, let --add-room also use them

  [ Jason Rhinelander ]
  * Option wording tweaks
  * Print default room permissions in room summary
  * Fix u/u2 code
  * Properly handle subrequest exceptions
  * Fix exception error conversion
  * Don't set empty `reactions` for deleted messages
  * Make blinding enabled by default

 -- Jason Rhinelander <jason@imaginary.ca>  Thu, 20 Oct 2022 18:05:44 -0300

session-pysogs (0.3.4-1) sid; urgency=medium

  [ Jason Rhinelander ]
  * Fix upgraded database not allowing deletion of reacted messages
  * Return new seqno from reaction-modifying endpoints

  [ Jeff ]
  * check post body and post username when checking profanity filter

  [ Jason Rhinelander ]
  * Return info_updates from room-modifying endpoints
  * Add "you" reactor threshold test case
  * Fix info_updates not updating on hidden global mod changes

 -- Jason Rhinelander <jason@imaginary.ca>  Wed, 24 Aug 2022 18:12:59 -0300

session-pysogs (0.3.3-1) sid; urgency=medium

  * Fix wrong reference in upgraded sqlite `files` table
  * Version bump

 -- Jason Rhinelander <jason@imaginary.ca>  Mon, 15 Aug 2022 02:46:43 -0300

session-pysogs (0.3.2-1) sid; urgency=medium

  * Bump dev version
  * Unicode onion request and auth fixes

 -- Jason Rhinelander <jason@imaginary.ca>  Thu, 11 Aug 2022 14:51:22 -0300

session-pysogs (0.3.1-1) sid; urgency=medium

  [ Jason Rhinelander ]
  * Fix missing packaged schema files

  [ Ian Macdonald ]
  * Make sogs also display the number of users active in the last day.

  [ Jason Rhinelander ]
  * Override active prune threshold to 1000 years in upgrade tests
  * Remove non-working cleanup
  * Doc generation: Omit the static handler endpoint
  * Remove legacy docs
  * Improve doc gen error message
  * Un-DRY query parameter docstrings
  * Remove done TODO
  * Document public rooms
  * Convert generation to use docsify
  * Catch exception when blinding fails
  * Don't try blinding already-blinded users
  * Exit after a no-such-room error
  * Avoid PIL deprecation warning
  * Quote "user" and other sync ups with postgresql schema
  * Reactions
  * Add reaction tests
  * Detect missing tables in pg import script

  [ Audric Ackermann ]
  * make ban/unban and mod endpoint accept blinded or not id

  [ Jason Rhinelander ]
  * add reactions to pg import
  * Add reactions capability
  * Fix docs to match what is actually returned
  * Include `reactions` key even when there are no reactions
  * Reactions ordering
  * Add "deleted": true when a message is deleted
  * Bump stable version

 -- Jason Rhinelander <jason@imaginary.ca>  Mon, 08 Aug 2022 21:18:10 -0300

session-pysogs (0.3.0-2) sid; urgency=medium

  * Fix missing packaged schema files

 -- Jason Rhinelander <jason@imaginary.ca>  Tue, 12 Apr 2022 22:23:04 -0300

session-pysogs (0.3.0-1) sid; urgency=medium

  * Really fix deletion failure
  * Fix python3-nacl version dependency
  * Bump version to 0.3.0 final

 -- Jason Rhinelander <jason@imaginary.ca>  Wed, 30 Mar 2022 18:35:32 -0300

session-pysogs (0.3.0~rc1-1) sid; urgency=medium

  [ Jason Rhinelander ]
  * Hide system user from get_all_global_moderators
  * Use SQLAlchemy for database backend
  * PostgreSQL support
  * Remove obsolete rust format config
  * Fix deprecated call: logger.warn -> logger.warning
  * DB interface fixes/improvements

sogs/db.py
from . import config
from . import crypto
from .postfork import postfork
import os
import logging
import importlib.resources
import sqlalchemy
from sqlalchemy.sql.expression import bindparam

HAVE_FILE_ID_HACKS = False
# roomid => (max, offset).  Max is the highest message id that was in the old table; offset is the
# value we add to ids <= that max to calculate the new database message id.
ROOM_IMPORT_HACKS = {}


def get_conn():
    """Gets a connection from the database engine connection pool.  This is not intended to be used
    by flask endpoints: they should use web.appdb instead (which calls this upon first use)."""
    return engine.connect()


def query(query, *, dbconn=None, bind_expanding=None, **params):
    """Executes a query containing :param style placeholders (regardless of the actual underlying
    database placeholder style), binding them using the given params keyword arguments.

    Note that, if the query contains a literal : it must be escaped as \\:

    For example:

        rows = db.query(
            "SELECT * FROM table1 WHERE name = :name AND age >= :age",
            name="Joe",
            age=25)

    See sqlalchemy.text for details.

    bind_expanding can be passed a sequence of bind names that are "expanding" to a tuple, most
    commonly used to bind and expand the RHS of a `x IN :x` clause.

    Can execute on a specific connection by passing it as dbconn; if omitted, uses web.appdb.  (Note
    that dbconn *cannot* be used as a placeholder bind name).
    """

    if dbconn is None:
        from . import web

        dbconn = web.appdb

    q = sqlalchemy.text(query)

    if bind_expanding:
        q = q.bindparams(*(bindparam(c, expanding=True) for c in bind_expanding))

    return dbconn.execute(q, **params)


# Begins a (potentially nested) transaction.  Takes an optional connection; if omitted uses
# web.appdb.
def transaction(dbconn=None):
    if dbconn is None:
        from . import web

        dbconn = web.appdb
    return dbconn.begin_nested()


have_returning = True


def insert_and_get_pk(insert, _pk, *, dbconn=None, **params):
    """
    Performs an insert and returns the value of the primary key by appending a RETURNING clause, if
    supported, and otherwise falling back to using .lastrowid.

    Takes the query, primary key column name, and any parameters to bind

    Can optionally take the database connection by passing as a dbconn parameter (note that you may
    not use "dbconn" as a bind parameter).  If omitted uses web.appdb.
    """

    if have_returning:
        insert += f" RETURNING {_pk}"

    result = query(insert, dbconn=dbconn, **params)
    if have_returning:
        return result.first()[0]
    return result.lastrowid


def insert_and_get_row(insert, _table, _pk, *, dbconn=None, **params):
    """
    Performs an insert and returned the row by appending a `RETURNING *` clause, if supported, and
    otherwise fetching the row immediately after the insertion.

    Takes the query, table name, and primary key column name (the latter two are needed for the
    SELECT query when the db doesn't support RETURNING), and any parameters to bind.

    Can optionally take the database connection by passing as a dbconn parameter (note that you may
    not use "dbconn" as a bind parameter).  If omitted uses web.appdb.
    """

    if have_returning:
        insert += " RETURNING *"
        return query(insert, dbconn=dbconn, **params).first()

    with transaction(dbconn):
        pkval = insert_and_get_pk(insert, _pk, dbconn=dbconn, **params)
        return query(f"SELECT * FROM {_table} WHERE {_pk} = :pk", pk=pkval).first()


def database_init(create=None, upgrade=True):
    """
    Perform database initialization: constructs the schema, if necessary, and performs any required
    migrations.  This does so using its *own* database connection, and is intended to be called
    during initialization *before* forking happens during uwsgi startup.

    create -- if True then we require that the database tables not already exist and raise an error
    if they do; if False then we raise an error if the database looks empty.  If None/unspecified
    then we will automatically create tables if required.

    upgrade -- if True (or omitted) then we allow database upgrades to run, if False then we throw
    an exception if any database upgrade is required.

    Returns true if database creation or upgrades were performed.
    """

    global engine, metadata

    metadata.clear()
    metadata.reflect(bind=engine, views=True)

    conn = get_conn()

    created, migrated = False, False
    if 'messages' not in metadata.tables:
        if create is not None and not create:
            raise RuntimeError(
                "Empty database connection and database initialization disabled; aborting"
            )

        logging.warning("No database detected; creating new database schema")
        if engine.name == "sqlite":
            conn.connection.executescript(importlib.resources.read_text('sogs', 'schema.sqlite'))
        elif engine.name == "postgresql":
            cur = conn.connection.cursor()
            cur.execute(importlib.resources.read_text('sogs', 'schema.pgsql'))
            cur.close()
        else:
            err = f"Don't know how to create the database for {engine.name}"
            logging.critical(err)
            raise RuntimeError(err)

        created = True
        metadata.clear()
        metadata.reflect(bind=engine, views=True)

        if 'messages' not in metadata.tables:
            msg = (
                "Critical error: SQL schema creation failed; "
                f"tables: {', '.join(metadata.tables.keys())}"
            )
            logging.critical(msg)
            raise RuntimeError(msg)
    elif create:
        raise RuntimeError("Unable to initialize database: tables already exist")

    from . import migrations

    migrated = migrations.migrate(conn, check_only=not upgrade and not created)
    if migrated:
        metadata.clear()
        metadata.reflect(bind=engine, views=True)

    # Make sure the system admin users exists
    create_admin_user(conn)

    check_needs_blinding(conn)

    return created or migrated


def create_admin_user(dbconn):
    """
    We create a dummy user (with id 0) for system tasks such as changing moderators from
    command-line, and give it the server's x25519 pubkey (with ff prepended, *not* 05) as a fake
    default session_id.
    """
    query(
        """
        INSERT INTO users (id, session_id, moderator, admin, visible_mod)
            VALUES (0, :sid, TRUE, TRUE, FALSE)
        ON CONFLICT (id) DO UPDATE
            SET session_id = :sid, moderator = TRUE, admin = TRUE, visible_mod = FALSE
        """,
        sid="ff" + crypto.server_pubkey_hex,
        dbconn=dbconn,
    )


def check_needs_blinding(dbconn):
    if not config.REQUIRE_BLIND_KEYS:
        return

    with transaction(dbconn):
        for uid, sid in query(
            """
            SELECT id, session_id FROM users WHERE id IN (
                SELECT "user" FROM user_permission_overrides
                UNION
                SELECT "user" FROM user_permission_futures
                UNION
                SELECT "user" FROM user_ban_futures
                UNION
                SELECT id FROM users WHERE session_id LIKE '05%' AND (admin OR moderator OR banned)
                EXCEPT
                SELECT "user" FROM needs_blinding
            )
            AND session_id LIKE '05%'
            """,
            dbconn=dbconn,
        ):
            try:
                pos_derived = crypto.compute_blinded_abs_id(sid)
            except Exception as e:
                logging.warning(f"Failed to blind session_id {sid}: {e}")
                continue

            query(
                'INSERT INTO needs_blinding (blinded_abs, "user") VALUES (:blinded, :uid)',
                blinded=pos_derived,
                uid=uid,
                dbconn=dbconn,
            )


engine, engine_initial_pid, metadata = None, None, None


def init_engine(*args, **kwargs):
    """
    Initializes db.engine.  This is called automatically during import of this submodule when
    running as a uwsgi app, but needs to be called manually otherwise before any database-using code
    is invoked.

    This can also be called to reinitialize db.engine, but that usage should be confined to the test
    suite.

    Keyword arguments:
    sogs_preinit - a callable to invoke after setting up `engine` but before calling
    `database_init()`.
    sogs_skip_init - optional boolean; if specified and True then database_init() will not be
    called; the caller must call it manually before any other database functions are used.
    """
    global engine, engine_initial_pid, metadata, have_returning

    if engine is not None:
        engine.dispose()

    skip_init = kwargs.pop('sogs_skip_init', False)
    preinit = kwargs.pop('sogs_preinit', None)

    if not len(args) and not len(kwargs):
        if config.DB_URL == 'defer-init':
            return
        args = (config.DB_URL,)

    exec_opts_args = {}
    if args[0].startswith('postgresql'):
        exec_opts_args['isolation_level'] = 'READ COMMITTED'
    else:
        # SQLite's Python code is seriously broken, so we have to force off autocommit mode and turn
        # on driver-level autocommit (which we do below).
        exec_opts_args['autocommit'] = False

    engine = sqlalchemy.create_engine(*args, **kwargs).execution_options(**exec_opts_args)
    engine_initial_pid = os.getpid()
    metadata = sqlalchemy.MetaData()

    if engine.name == "sqlite":
        import sqlite3

        if sqlite3.sqlite_version_info < (3, 25, 0):
            raise RuntimeError(
                f"SQLite3 library version {'.'.join(sqlite3.sqlite_version_info)} "
                "is too old for pysogs (3.25.0+ required)!"
            )

        have_returning = sqlite3.sqlite_version_info >= (3, 35, 0)

        @sqlalchemy.event.listens_for(engine, "connect")
        def sqlite_fix_connect(dbapi_connection, connection_record):
            # disable pysqlite's emitting of the BEGIN statement entirely.
            # also stops it from emitting COMMIT before any DDL.
            dbapi_connection.isolation_level = None
            # Enforce foreign keys.  It is very sad that this is not default.
            cursor = dbapi_connection.cursor()
            cursor.execute("PRAGMA foreign_keys=ON")
            cursor.close()

        @sqlalchemy.event.listens_for(engine, "begin")
        def do_begin(conn):
            # emit our own BEGIN
            conn.execute("BEGIN IMMEDIATE")

    else:
        have_returning = True

        # rooms.token is a 'citext' (case-insensitive text), which sqlalchemy doesn't recognize out
        # of the box.  Map it to a plain TEXT which is good enough for what we need (if we actually
        # needed to generate this wouldn't suffice: we'd have to use something like the
        # sqlalchemy-citext module).
        from sqlalchemy.dialects.postgresql.base import ischema_names

        if 'citext' not in ischema_names:
            ischema_names['citext'] = ischema_names['text']

    if preinit:
        preinit()

    if not skip_init:
        database_init()


if config.RUNNING_AS_APP:
    init_engine()


@postfork
def reset_db_postfork():
    """Clear any connections from the engine after forking because they aren't shareable."""
    if engine is None or os.getpid() == engine_initial_pid:
        return
    engine.dispose()
